testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestReadUniontype.java

```diff
@@ -139,6 +137,127 @@ public void testReadUniontype(String storageFormat)
+             *  at org.apache.hadoop.hive.ql.exec.mr.ExecMapper.map(ExecMapper.java:170)
+             *  ... 8 more
+             * Caused by: java.lang.ArrayIndexOutOfBoundsException: 1
+             *  at org.apache.avro.generic.GenericData$Record.get(GenericData.java:135)
+             *  at org.apache.avro.generic.GenericData.getField(GenericData.java:580)
+             *  at org.apache.avro.generic.GenericData.validate(GenericData.java:373)
+             *  at org.apache.avro.generic.GenericData.validate(GenericData.java:395)
+             *  at org.apache.avro.generic.GenericData.validate(GenericData.java:373)
+             *  at org.apache.hadoop.hive.serde2.avro.AvroSerializer.serialize(AvroSerializer.java:96)
+             *
+             * So we try coercion logic on the first struct field only.
+             *
+             */
+
+            // Generate a file with rows:
+            //   0, {0: <a="a1",b="b1">}
+            //   1, {1: <c="c1">}
+            onHive().executeQuery(format(
+                    "INSERT INTO TABLE %s PARTITION (c2 = 5) "
+                            + "SELECT 0, create_union(0, named_struct('a', 'a1', 'b', 'b1'), named_struct('c', 'ignore')) "
+                            + "UNION ALL "
+                            + "SELECT 1, create_union(0, named_struct('a', 'a2', 'b', 'b2'), named_struct('c', 'ignore'))",
+                    TABLE_NAME_SCHEMA_EVOLUTION));
+
+            // Add a coercible change inside union type column.
+            onHive().executeQuery(format("ALTER TABLE %S CHANGE COLUMN c1 c1 UNIONTYPE<STRUCT<a:STRING, b:STRING, d:STRING>, STRUCT<c:STRING>>", TABLE_NAME_SCHEMA_EVOLUTION));
+
+            QueryResult selectAllResult = onTrino().executeQuery(format("SELECT c0, c1 FROM %s", TABLE_NAME_SCHEMA_EVOLUTION));
+            assertEquals(selectAllResult.rows().size(), 2);
+            for (List<?> row : selectAllResult.rows()) {
+                int id = (Integer) row.get(0);
+                switch (id) {
+                    case 0:
+                        Row rowValueFirst = rowBuilder()
+                                .addField("a", "a1")
+                                .addField("b", "b1")
+                                .addField("d", null)
+                                .build();
+                        assertStructEquals(row.get(1), new Object[] {(byte) 0, rowValueFirst, null});
+                        break;
+                    case 1:
+                        Row rowValueSecond = rowBuilder()
+                                .addField("a", "a2")
+                                .addField("b", "b2")
+                                .addField("d", null)
+                                .build();
+                        assertStructEquals(row.get(1), new Object[] {(byte) 0, rowValueSecond, null});
+                        break;
+                }
+            }
+        }
+        else {
```

Comment:
let's refactor this a bit to have different delegate methods for avro and orc. there doesn't seem to be any shared code after the table creation